### ISSUE TITLE
Storage migration fails, causes subsequent start to fail

### ISSUE THREAD

**Issue by stgraber** [lxc/lxd#2907]
```
Jan 31 21:00:58 athos systemd[1]: Starting LXD - main daemon...
Jan 31 21:00:58 athos lxd[8713]: lvl=warn msg="CGroup memory swap accounting is disabled, swap limits will be ignored." t=2017-01-31T21:00:58+0000
Jan 31 21:00:58 athos lxd[8713]: lvl=warn msg="Unable to refresh cache, using stale entry" server=https://cloud-images.ubuntu.com/releases t=2017-01-31T21:00:5
Jan 31 21:01:04 athos systemd[1]: Started LXD - main daemon.
Feb 18 11:11:24 athos systemd[1]: Stopping LXD - main daemon...
Feb 18 11:11:24 athos systemd[1]: Stopped LXD - main daemon.
Feb 18 11:11:27 athos systemd[1]: Starting LXD - main daemon...
Feb 18 11:11:27 athos lxd[547145]: lvl=warn msg="CGroup memory swap accounting is disabled, swap limits will be ignored." t=2017-02-18T11:11:27+0000
Feb 18 11:11:43 athos lxd[547145]: error: Failed to unmount ZFS filesystem: umount: /var/lib/lxd/containers/libvirt01.zfs: target is busy
Feb 18 11:11:43 athos lxd[547145]:         (In some cases useful info about processes that
Feb 18 11:11:43 athos lxd[547145]:          use the device is found by lsof(8) or fuser(1).)
Feb 18 11:11:43 athos lxd[547145]: cannot unmount '/var/lib/lxd/containers/libvirt01.zfs': umount failed
Feb 18 11:11:43 athos systemd[1]: lxd.service: Main process exited, code=exited, status=1/FAILURE
Feb 18 11:21:27 athos lxd[547146]: error: LXD still not running after 600s timeout.
Feb 18 11:21:27 athos systemd[1]: lxd.service: Control process exited, code=exited status=1
Feb 18 11:21:27 athos systemd[1]: Failed to start LXD - main daemon.
Feb 18 11:21:27 athos systemd[1]: lxd.service: Unit entered failed state.
Feb 18 11:21:27 athos systemd[1]: lxd.service: Failed with result 'exit-code'.
Feb 18 11:21:27 athos systemd[1]: lxd.service: Service hold-off time over, scheduling restart.
Feb 18 11:21:27 athos systemd[1]: Stopped LXD - main daemon.
Feb 18 11:21:27 athos systemd[1]: Starting LXD - main daemon...
Feb 18 11:21:27 athos lxd[564449]: lvl=warn msg="CGroup memory swap accounting is disabled, swap limits will be ignored." t=2017-02-18T11:21:27+0000
Feb 18 11:21:28 athos lxd[564449]: error: UNIQUE constraint failed: storage_pools.name
Feb 18 11:21:28 athos systemd[1]: lxd.service: Main process exited, code=exited, status=1/FAILURE
Feb 18 11:31:27 athos lxd[564450]: error: LXD still not running after 600s timeout.
```

**Comment by stgraber** [comment]
```
root@athos:~# ls -lh /var/lib/lxd/
total 156K
drwxr-xr-x 1 root root  352 Feb 18 11:11 containers
drwx--x--x 1 root root  158 Jan 20 20:24 devices
drwxr-xr-x 1 root root    0 Feb 18 11:11 devlxd
drwx------ 1 root root    0 Feb 18 11:11 disks
drwx------ 1 root root 1.6K Feb  4 03:01 images
-rw-r--r-- 1 root root  75K Feb 18 11:11 lxd.db
-rw-r--r-- 1 root root  70K Feb 18 11:11 lxd.db.bak
drwx--x--x 1 root root    0 Jan 26 03:11 networks
drwx------ 1 root root   30 Dec 21  2015 security
-rw-r--r-- 1 root root 2.4K Dec 21  2015 server.crt
-rw------- 1 root root 3.2K Dec 21  2015 server.key
drwxr-xr-x 1 root root  158 Jan 20 20:24 shmounts
drwx------ 1 root root   18 Aug 29 03:31 snapshots
drwx--x--x 1 root root   14 Feb 18 11:11 storage-pools
srw-rw---- 1 root lxd     0 Jan 31 21:00 unix.socket

root@athos:~# ls -lh /var/lib/lxd/containers/
total 200K
lrwxrwxrwx 1 root    root     56 Feb 18 11:11 backuppc01 -> /var/lib/lxd/storage-pools/default/containers/backuppc01
lrwxrwxrwx 1 root    root     53 Feb 18 11:11 cacti01 -> /var/lib/lxd/storage-pools/default/containers/cacti01
lrwxrwxrwx 1 root    root     55 Feb 18 11:11 jenkins01 -> /var/lib/lxd/storage-pools/default/containers/jenkins01
lrwxrwxrwx 1 root    root     37 Dec 21  2015 libvirt01 -> /var/lib/lxd/containers/libvirt01.zfs
drwx------ 5 root    root      7 Jan 20 04:35 libvirt01.zfs
-rw-r--r-- 1 root    root    43K Jan 31 21:01 lxc-monitord.log
lrwxrwxrwx 1 root    root     35 Jan 20 20:22 lxd-kpi -> /var/lib/lxd/containers/lxd-kpi.zfs
drwxr-xr-x 4 1000000 1000000   6 Jan 10 14:35 lxd-kpi.zfs
lrwxrwxrwx 1 root    root     34 Feb 13  2016 maas01 -> /var/lib/lxd/containers/maas01.zfs
drwx------ 4 root    root      6 Jan 20 04:34 maas01.zfs
lrwxrwxrwx 1 root    root     35 Dec 21  2015 munin01 -> /var/lib/lxd/containers/munin01.zfs
drwxr-xr-x 4 1000000 1000000   6 Jan 20 04:33 munin01.zfs
lrwxrwxrwx 1 root    root     36 Dec 21  2015 nagios01 -> /var/lib/lxd/containers/nagios01.zfs
drwxr-xr-x 4 1000000 1000000   6 Jan 20 04:33 nagios01.zfs
lrwxrwxrwx 1 root    root     34 Jun 14  2016 plex01 -> /var/lib/lxd/containers/plex01.zfs
drwxr-xr-x 4 1000000 1000000   6 Jan 20 04:33 plex01.zfs
lrwxrwxrwx 1 root    root     38 Dec 21  2015 samba-fs01 -> /var/lib/lxd/containers/samba-fs01.zfs
drwxr-xr-x 5 1000000 1000000   7 Jan 20 04:33 samba-fs01.zfs

root@athos:~# ls -lh /var/lib/lxd/disks/
total 0

root@athos:~# find /var/lib/lxd/storage-pools/
/var/lib/lxd/storage-pools/
/var/lib/lxd/storage-pools/default
/var/lib/lxd/storage-pools/default/containers
/var/lib/lxd/storage-pools/default/containers/backuppc01
/var/lib/lxd/storage-pools/default/containers/cacti01
/var/lib/lxd/storage-pools/default/containers/jenkins01

root@athos:~# zfs list -t all | grep containers
internal/lxd/containers                                                                                34.6G  35.5G   192K  none
internal/lxd/containers/backuppc01                                                                      549M  35.5G   678M  /var/lib/lxd/storage-pools/default/containers/backuppc01
internal/lxd/containers/cacti01                                                                         606M  35.5G   735M  /var/lib/lxd/storage-pools/default/containers/cacti01
internal/lxd/containers/jenkins01                                                                      1021M  35.5G  1.16G  /var/lib/lxd/storage-pools/default/containers/jenkins01
internal/lxd/containers/libvirt01                                                                      1.11G  35.5G   641M  /var/lib/lxd/containers/libvirt01.zfs
internal/lxd/containers/libvirt01@snapshot-pre-upgrade                                                  505M      -   776M  -
internal/lxd/containers/lxd-kpi                                                                        1.48G  35.5G  1.48G  /var/lib/lxd/containers/lxd-kpi.zfs
internal/lxd/containers/maas01                                                                         20.5G  35.5G  20.5G  /var/lib/lxd/containers/maas01.zfs
internal/lxd/containers/munin01                                                                        5.84G  35.5G  5.96G  /var/lib/lxd/containers/munin01.zfs
internal/lxd/containers/nagios01                                                                       1.13G  35.5G  1.26G  /var/lib/lxd/containers/nagios01.zfs
internal/lxd/containers/plex01                                                                         1.93G  35.5G  2.14G  /var/lib/lxd/containers/plex01.zfs
internal/lxd/containers/samba-fs01                                                                      510M  35.5G   634M  /var/lib/lxd/containers/samba-fs01.zfs
```

**Comment by stgraber** [comment]
```
sqlite> SELECT * FROM config;
1|storage.zfs_pool_name|internal/lxd
4|core.https_address|[::]


sqlite> SELECT * FROM storage_pools;
1|default|zfs

sqlite> SELECT * FROM storage_volumes;
1|backuppc01|1|0
2|cacti01|1|0
3|jenkins01|1|0
4|libvirt01|1|0
```

**Comment by stgraber** [comment]
```
root@athos:~# lxd --debug --group lxd
INFO[02-19|06:40:06] LXD 2.9.1 is starting in normal mode     path=/var/lib/lxd
WARN[02-19|06:40:06] CGroup memory swap accounting is disabled, swap limits will be ignored. 
INFO[02-19|06:40:06] Default uid/gid map: 
INFO[02-19|06:40:06]  - u 0 1000000 1000000 
INFO[02-19|06:40:06]  - g 0 1000000 1000000 
DBUG[02-19|06:40:06] Initializing and checking storage pool "default". 
DBUG[02-19|06:40:06] Treating the source "internal/lxd" of this storage pool as a ZFS dataset. driver=storage/zfs
DBUG[02-19|06:40:06] StoragePoolInit                          driver=storage/zfs
DBUG[02-19|06:40:06] StoragePoolCheck                         driver=storage/zfs
DBUG[02-19|06:40:06] Applying patch: storage_api 
error: UNIQUE constraint failed: storage_pools.name
```

**Comment by stgraber** [comment]
```
root@athos:~# cp /var/lib/lxd/lxd.db /var/lib/lxd/lxd.db.broken
root@athos:~# cp /var/lib/lxd/lxd.db.bak /var/lib/lxd/lxd.db
```

 - Called "umount -l" for every entry in /var/lib/lxd/containers/
 - Made sure that each container had both a /var/lib/lxd/containers/NAME and /var/lib/lxd/containers/NAME.zfs entry (just touching an empty file if missing)
 - Ran "lxd --debug --group lxd" again and migration succeeded.

**Comment by stgraber** [comment]
The conclusion is that the migration code must be robust enough to survive if the ZFS filesystem refuses to be unmounted by normal means (calling umount -l on all of them seems like a potential way around this issue).

**Comment by stgraber** [comment]
#2906 also showed that a similar failure to unmount an image can lead to the same problem. We should also lazily unmount images.

**Comment by stgraber** [comment]
This issue should now have been resolved in git and in LXD 2.9.2, closing.

If anyone who has installed LXD 2.9.2 runs into something like this, please file another issue.
